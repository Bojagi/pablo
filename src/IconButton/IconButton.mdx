---
name: IconButton
route: /iconbutton
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { IconButton } from './IconButton.tsx'
import { Trash2, Italic, CheckCircle, Crop } from 'react-feather';

# IconButton

IconButtons are buttons that only consists of an icon. They are useful for actions on places where space is rare
and an icon on it's own is really descriptive.

## Examples

### Sized IconButton

You can display `IconButton`s in different sizes. By default they have the size `medium`. 
They can also be `small` and `large`. 

<Playground>
  <Box display="flex" mx={-4}>
    <IconButton mx={4} onClick="You clicked me!">
      <Trash2 />
    </IconButton>
    <IconButton size="small" mx={4} onClick="You clicked me!">
      <Trash2 />
    </IconButton>
    <IconButton size="large" mx={4} onClick="You clicked me!">
      <Trash2 />
    </IconButton>
  </Box>
</Playground>

### Active Icon buttons
IconButtons can also be used to toggle certain functionality (e.g. bold text mode in rich text inputs). For this
you can use the boolean `active` prop.

<Playground>
  {() => {
    const [active, setActive] = React.useState(props.active);
    return (
      <IconButton active={active} onClick={() => setActive(!active)}>
        <Italic />
      </IconButton>
    );;
  }}
</Playground>

### Colored IconButton

To give IconButtons a different color, just use the `color` prop. Colors can be `brand`, `plain`, `negative` or `positive`.
The default color is `plain`. 

<Playground>
  {() => {
    const [active, setActive] = React.useState(props.active);
    return (
      <Box display="flex" mx={-4}>
        <IconButton mx={4} color="brand" active={active} onClick={() => setActive(!active)}>
          <Crop />
        </IconButton>
        <IconButton mx={4} color="positive" active={active} onClick={() => setActive(!active)}>
          <CheckCircle />
        </IconButton>
        <IconButton mx={4} color="negative" active={active} onClick={() => setActive(!active)}>
          <Trash2 />
        </IconButton>
      </Box>
    );
  }}
</Playground>

## Import

```js
import { IconButton } from '@bojagi/pablo/IconButton';
```

## Props

<Props of={IconButton} />
