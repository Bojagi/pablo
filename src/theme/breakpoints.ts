export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
