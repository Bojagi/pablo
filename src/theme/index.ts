export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
