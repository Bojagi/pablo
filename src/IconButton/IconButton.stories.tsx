import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Check } from 'react-feather';
import { css } from 'styled-components';
import { Flex } from '../Box';
import { Subtitle } from '../Typography';
import { IconButton } from './IconButton';

export default {
  title: 'IconButton',
};

const ControlledIconButton = (props) => {
  const [active, setActive] = React.useState(props.active);
  return <IconButton {...props} active={active} onClick={() => setActive(!active)} />;
};

const baseStory = (args) => (
  <>
    <Subtitle>Small</Subtitle>
    <Flex mx={-4} mb={5}>
      <ControlledIconButton mx={4} size="small" {...args}>
        <Check />
      </ControlledIconButton>
      <ControlledIconButton mx={4} size="small" disabled {...args}>
        <Check />
      </ControlledIconButton>
    </Flex>
    <Subtitle>Medium</Subtitle>
    <Flex mx={-4} mb={5}>
      <ControlledIconButton mx={4} size="medium" {...args}>
        <Check />
      </ControlledIconButton>
      <ControlledIconButton mx={4} size="medium" disabled {...args}>
        <Check />
      </ControlledIconButton>
    </Flex>
    <Subtitle>Large</Subtitle>
    <Flex mx={-4} mb={5}>
      <ControlledIconButton mx={4} size="large" {...args}>
        <Check />
      </ControlledIconButton>
      <ControlledIconButton mx={4} size="large" disabled {...args}>
        <Check />
      </ControlledIconButton>
    </Flex>
  </>
);

export const Normal = baseStory.bind(null);
Normal.args = {};

export const Active = baseStory.bind(null);
Active.args = {
  active: true,
};

export const CustomStyles = baseStory.bind(null);
CustomStyles.args = {
  customStyles: {
    root: css`
      background-color: red;
    `,
    focus: css`
      background-color: blue;
    `,
    hover: css`
      background-color: darkred;
    `,
    active: css`
      background-color: yellow;
    `,
  },
};
