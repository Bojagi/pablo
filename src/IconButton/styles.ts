import { Style } from '../theme/types';
import { getColor, getSpacing } from '../utils/styleHelpers';

export interface IconButtonStyles {
  size: {
    small: Style;
    medium: Style;
    large: Style;
  };
  icon: {
    scale: number;
    transition: string[][];
    active: {
      scale: number;
    };
    size: {
      small: Style;
      medium: Style;
      large: Style;
    };
  };
  borderRadius: number;
  backgroundColor: Style;
  color: Style;
  transition: string[][];
  active: {
    backgroundColor: Style;
    color: Style;
  };
  disabled: {
    opacity: number;
  };
  hover: {
    backgroundColor: Style;
    color: Style;
  };
  focus: {
    backgroundColor: Style;
    color: Style;
  };
}

export const iconButtonStyles: IconButtonStyles = {
  size: {
    small: getSpacing(3),
    medium: getSpacing(4),
    large: getSpacing(5),
  },
  icon: {
    scale: 1,
    transition: [['transform', '0.2s']],
    active: {
      scale: 0.833333333,
    },
    size: {
      small: getSpacing(2.5),
      medium: getSpacing(3.5),
      large: getSpacing(4.5),
    },
  },
  borderRadius: 6,
  backgroundColor: 'transparent',
  color: getColor('common', 'black'),
  transition: [
    ['background-color', '0.2s'],
    ['color', '0.2s'],
  ],
  active: {
    backgroundColor: getColor('brand'),
    color: getColor('common', 'white'),
  },
  disabled: {
    opacity: 0.3,
  },
  hover: {
    backgroundColor: getColor('gray', '50'),
    color: getColor('common', 'black'),
  },
  focus: {
    backgroundColor: getColor('gray', '100'),
    color: getColor('common', 'black'),
  },
};
