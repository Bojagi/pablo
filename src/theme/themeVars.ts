import isObject from 'isobject';
import { breakpointVars } from './breakpoints';
import { colorVars } from './colors';
import { typographyVars } from './typography';

export const themeVarNames = {
  typography: typographyVars,
  colors: colorVars,
  breakpoints: breakpointVars,
};

function createThemeReferences<T extends Record<string, any>>(obj: T): T {
  return Object.entries(obj).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]: isObject(value) ? createThemeReferences(value as T) : `var(--${value})`,
    }),
    {} as any
  );
}

export const themeVars = createThemeReferences(themeVarNames);
