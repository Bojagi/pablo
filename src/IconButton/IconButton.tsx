import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { LayoutBoxProps, layoutInterpolationFn } from '../Box';
import { baseStyle } from '../shared/baseStyle';
import { getComponentStyle, transitionTransformer } from '../styleHelpers';
import { BaseProps, CssFunctionReturn } from '../types';
import { getCustomStyles } from '../utils/useCustomStyles';
import { IconButtonStyleProperties } from './styles';
import { omit } from '../utils/omit';
import React, { forwardRef, HTMLProps } from 'react';

export type IconButtonSize = 'small' | 'medium' | 'large';
export type IconButtonColor = 'brand' | 'plain' | 'negative' | 'positive';

export interface IconButtonProps
  extends Omit<LayoutBoxProps, 'theme'>,
    BaseProps<IconButtonStyleProperties>,
    Omit<HTMLProps<HTMLButtonElement>, 'width' | 'height' | 'size' | 'as' | 'type'> {
  size?: IconButtonSize;
  color?: IconButtonColor;
  active?: boolean;
  role?: string;
  children?: React.ReactNode;
  css?: CssFunctionReturn<IconButtonProps>;
}
const StyledIconButton = styled.button<IconButtonProps>`
  ${baseStyle}
  width: ${getComponentStyle('iconButton.size.{size}')};
  height: ${getComponentStyle('iconButton.size.{size}')};
  background-color: ${getComponentStyle('iconButton.{color}.backgroundColor')};
  color: ${getComponentStyle('iconButton.{color}.color')};

  border: 0;
  padding: 0;
  outline: 0;
  box-sizing: border-box;
  transition: ${getComponentStyle('iconButton.transition', transitionTransformer)};
  border-radius: ${getComponentStyle('iconButton.borderRadius')};
  display: flex;
  justify-content: center;
  align-items: center;
  flex-shrink: 0;

  & > * {
    width: ${getComponentStyle('iconButton.icon.size.{size}')};
    height: ${getComponentStyle('iconButton.icon.size.{size}')};
    transition: ${getComponentStyle('iconButton.icon.transition', transitionTransformer)};
    transform: scale(
      ${(props) =>
        props.active
          ? props.theme.componentStyles.iconButton.icon.active.scale
          : props.theme.componentStyles.iconButton.icon.scale}
    );
  }

  &:enabled {
    cursor: pointer;
  }

  &:disabled {
    opacity: ${getComponentStyle('iconButton.disabled.opacity')};
    cursor: normal;
  }

  ${getCustomStyles('iconButton.styles', 'root')}

  ${(props) =>
    props.active
      ? css`
          background-color: ${getComponentStyle('iconButton.{color}.active.backgroundColor')(
            props
          )};
          color: ${getComponentStyle('iconButton.{color}.active.color')(props)};
          ${getCustomStyles('iconButton.styles', 'active')(props)}
        `
      : css`
          &:hover:enabled {
            background-color: ${getComponentStyle('iconButton.{color}.hover.backgroundColor')(
              props
            )};
            color: ${getComponentStyle('iconButton.{color}.hover.color')(props)};
            ${getCustomStyles('iconButton.styles', 'hover')(props)}
          }

          &:focus:enabled {
            background-color: ${getComponentStyle('iconButton.{color}.focus.backgroundColor')(
              props
            )};
            color: ${getComponentStyle('iconButton.{color}.focus.color')(props)};
            ${getCustomStyles('iconButton.styles', 'focus')(props)}
          }
        `}

  ${(props) => layoutInterpolationFn(omit(props, ['size']))}
`;

const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  ({ size = 'medium', color = 'plain', active = false, ...props }, ref) => (
    <StyledIconButton
      size={size}
      color={color}
      active={active}
      ref={ref}
      aria-pressed={active ? 'true' : 'false'}
      {...props}
    />
  )
);

export { IconButton };
