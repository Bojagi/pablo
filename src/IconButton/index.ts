import './styles';

export * from './IconButton';
