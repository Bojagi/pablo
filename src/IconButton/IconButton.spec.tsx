import { render } from '@testing-library/react';
import React from 'react';
import { IconButton } from './IconButton';
import { PabloThemeProvider } from '../theme';
import { themeVars } from '../theme/themeVars';

test.each([
  ['small', '24px', '20px'],
  ['medium', '32px', '28px'],
  ['large', '40px', '36px'],
])(
  'Render %s icon button with correct styles',
  (sizeName, expectedButtonSize, expectedIconSize) => {
    const { container } = renderComponent({ size: sizeName });
    const buttonNode = container.firstChild;
    expect(buttonNode).toMatchSnapshot();
    // Button
    expect(buttonNode).toHaveStyleRule('width', expectedButtonSize);
    expect(buttonNode).toHaveStyleRule('height', expectedButtonSize);
    expect(buttonNode).toHaveStyleRule('background-color', 'transparent');
    expect(buttonNode).toHaveStyleRule('color', themeVars.colors.common.black);

    expect(buttonNode).toHaveStyleRule('background-color', themeVars.colors.blackOpacity['50'], {
      target: ':hover:enabled',
    });
    expect(buttonNode).toHaveStyleRule('color', themeVars.colors.common.black, {
      target: ':hover:enabled',
    });

    expect(buttonNode).toHaveStyleRule('background-color', themeVars.colors.blackOpacity['100'], {
      target: ':focus:enabled',
    });
    expect(buttonNode).toHaveStyleRule('color', themeVars.colors.common.black, {
      target: ':focus:enabled',
    });

    // Icon
    expect(buttonNode).toHaveStyleRule('transform', /scale\(\s*1\s*\)/, {
      target: '>*',
    });
    expect(buttonNode).toHaveStyleRule('width', expectedIconSize, {
      target: '>*',
    });
    expect(buttonNode).toHaveStyleRule('height', expectedIconSize, {
      target: '>*',
    });
  }
);

test('Render active with correct styles', () => {
  const { container } = renderComponent({ active: true });
  const buttonNode = container.firstChild;
  // Button
  expect(buttonNode).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(buttonNode).toHaveStyleRule('color', themeVars.colors.common.white);

  // Icon
  expect(buttonNode).toHaveStyleRule('transform', /scale\(\s*0.833333333\s*\)/, {
    target: '>*',
  });
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <IconButton {...props}>
        <div>Icon</div>
      </IconButton>
    </PabloThemeProvider>
  );
}
