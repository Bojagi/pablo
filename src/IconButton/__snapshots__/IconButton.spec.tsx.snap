// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render large icon button with correct styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
  background-color: transparent;
  color: var(--pbl-theme-colors-common-black);
  border: 0;
  padding: 0;
  outline: 0;
  box-sizing: border-box;
  -webkit-transition: background-color 0.2s,color 0.2s;
  transition: background-color 0.2s,color 0.2s;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0>* {
  width: 36px;
  height: 36px;
  -webkit-transition: -webkit-transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(
        1
      );
  -moz-transform: scale(
        1
      );
  -ms-transform: scale(
        1
      );
  transform: scale(
        1
      );
}

.emotion-0:enabled {
  cursor: pointer;
}

.emotion-0:disabled {
  opacity: 0.3;
  cursor: normal;
}

.emotion-0:hover:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-50);
  color: var(--pbl-theme-colors-common-black);
}

.emotion-0:focus:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-100);
  color: var(--pbl-theme-colors-common-black);
}

<button
  aria-pressed="false"
  class="emotion-0"
  color="plain"
>
  <div>
    Icon
  </div>
</button>
`;

exports[`Render medium icon button with correct styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
  background-color: transparent;
  color: var(--pbl-theme-colors-common-black);
  border: 0;
  padding: 0;
  outline: 0;
  box-sizing: border-box;
  -webkit-transition: background-color 0.2s,color 0.2s;
  transition: background-color 0.2s,color 0.2s;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0>* {
  width: 28px;
  height: 28px;
  -webkit-transition: -webkit-transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(
        1
      );
  -moz-transform: scale(
        1
      );
  -ms-transform: scale(
        1
      );
  transform: scale(
        1
      );
}

.emotion-0:enabled {
  cursor: pointer;
}

.emotion-0:disabled {
  opacity: 0.3;
  cursor: normal;
}

.emotion-0:hover:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-50);
  color: var(--pbl-theme-colors-common-black);
}

.emotion-0:focus:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-100);
  color: var(--pbl-theme-colors-common-black);
}

<button
  aria-pressed="false"
  class="emotion-0"
  color="plain"
>
  <div>
    Icon
  </div>
</button>
`;

exports[`Render small icon button with correct styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  width: 24px;
  height: 24px;
  background-color: transparent;
  color: var(--pbl-theme-colors-common-black);
  border: 0;
  padding: 0;
  outline: 0;
  box-sizing: border-box;
  -webkit-transition: background-color 0.2s,color 0.2s;
  transition: background-color 0.2s,color 0.2s;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0>* {
  width: 20px;
  height: 20px;
  -webkit-transition: -webkit-transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(
        1
      );
  -moz-transform: scale(
        1
      );
  -ms-transform: scale(
        1
      );
  transform: scale(
        1
      );
}

.emotion-0:enabled {
  cursor: pointer;
}

.emotion-0:disabled {
  opacity: 0.3;
  cursor: normal;
}

.emotion-0:hover:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-50);
  color: var(--pbl-theme-colors-common-black);
}

.emotion-0:focus:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-100);
  color: var(--pbl-theme-colors-common-black);
}

<button
  aria-pressed="false"
  class="emotion-0"
  color="plain"
>
  <div>
    Icon
  </div>
</button>
`;
